File: cpp/examples/payload_with_image_example.cc
#include <drift_protocol/common/data_payload.pb.h>
#include <drift_protocol/common/drift_package.pb.h>
#include <google/protobuf/util/time_util.h>

#include <wavelet_buffer/wavelet_buffer.h>
#include <wavelet_buffer/img/wavelet_image.h>
#include <wavelet_buffer/img/jpeg_codecs.h>
#include <wavelet_buffer/denoise_algorithms.h>


using drift::proto::common::DataPayload;
using drift::proto::common::DriftPackage;
using drift::proto::common::StatusCode;
using drift::proto::meta::MetaInfo;
using drift::proto::meta::ImageInfo;

using drift::WaveletBuffer;
using drift::WaveletParameters;
using drift::WaveletTypes;
using drift::img::WaveletImage;
using Denoiser = drift::SimpleDenoiseAlgorithm<float>;
using drift::img::RgbJpegCodec;

using google::protobuf::util::TimeUtil;
using google::protobuf::Any;

int main() {
    const auto pb_time = TimeUtil::GetCurrentTime();
    std::string message;

    // Load and decompose image with wavelet image
    const auto kWidth = 800;
    const auto kHeight = 500;
    WaveletImage img(WaveletParameters{
            .signal_shape = {kWidth, kHeight},
            .signal_number = 3,
            .decomposition_steps = 3,
            .wavelet_type = WaveletTypes::kDB3,
    });

    // Load JPG image, decode, decompose with wavelet transformation and denoise
    if (auto code = img.ImportFromFile(IMAGE_PATH, Denoiser(0.9), RgbJpegCodec())) {
        std::cerr << "Failed to import image " << code;
        return -1;
    }

    {
        // Create a package and serialize it
        DriftPackage original;
        original.set_id(TimeUtil::TimestampToMilliseconds(pb_time)); // UNIX timestamp in ms
        original.set_status(StatusCode::GOOD);
        original.mutable_source_timestamp()->CopyFrom(pb_time);
        original.mutable_publish_timestamp()->CopyFrom(pb_time);

        // Fill meta data
        ImageInfo image_info;
        image_info.set_type(ImageInfo::WB); // this is image decode with WaveletBuffer
        image_info.set_width(kWidth);
        image_info.set_height(kHeight);
        image_info.set_channel_layout("RGB");

        MetaInfo meta;
        meta.set_type(MetaInfo::IMAGE);
        meta.mutable_image_info()->CopyFrom(image_info);

        original.mutable_meta()->CopyFrom(meta);

        // Prepare payload with the image
        std::string compressed_image;
        if (!img.buffer().Serialize(&compressed_image, 16)) {
            std::cerr << "Failed to compress image image";
            return -1;
        }

        DataPayload payload;
        payload.set_data(compressed_image);
        original.add_data()->PackFrom(payload);

        // Serialize package to message
        original.SerializePartialToString(&message);
    }

    // Parse the package
    DriftPackage new_package;
    new_package.ParseFromString(message);
    std::cout << "Package with ID=" << new_package.id() << " type="
              << MetaInfo::DataType_Name(new_package.meta().type()) << std::endl;

    DataPayload payload;
    new_package.data(0).UnpackTo(&payload);
    auto buffer = WaveletBuffer::Parse(payload.data());
    std::cout << "Wavelet Buffer: " << buffer->parameters() << std::endl;
}

File: python/examples/payload_with_image_example.py
from pathlib import Path

from drift_protocol.common import DriftPackage, DataPayload, StatusCode
from drift_protocol.meta import MetaInfo, ImageInfo
from wavelet_buffer import denoise, WaveletType, WaveletBuffer
from wavelet_buffer.img import WaveletImage, RgbJpeg

from google.protobuf.timestamp_pb2 import Timestamp
from google.protobuf.any_pb2 import Any

IMG_PATH = Path(__file__).parent / "pandas.jpg"

if __name__ == "__main__":
    pb_time = Timestamp()
    pb_time.GetCurrentTime()

    # Load and decompose image with wavelet image
    WIDTH = 800
    HEIGHT = 500
    img = WaveletImage(signal_shape=[WIDTH, HEIGHT], signal_number=3, decomposition_steps=3,
                       wavelet_type=WaveletType.DB3)

    img.import_from_file(str(IMG_PATH.resolve()), denoiser=denoise.Simple(0.9), codec=RgbJpeg())

    # Create a package and serialize it
    original = DriftPackage()
    original.id = pb_time.ToMilliseconds()
    original.source_timestamp.CopyFrom(pb_time)
    original.publish_timestamp.CopyFrom(pb_time)
    original.status = StatusCode.GOOD

    # Fill meta data
    image_info = ImageInfo()
    image_info.type = ImageInfo.ImageType.WB
    image_info.width = WIDTH
    image_info.height = HEIGHT
    image_info.channel_layout = "RGB"

    meta = MetaInfo
    meta.type = MetaInfo.IMAGE
    meta.image_info  = image_info

    # Prepare payload
    payload = DataPayload()
    payload.data = img.buffer.serialize(compression_level=16)

    msg = Any()
    msg.Pack(payload)
    original.data.append(msg)

    # Serialize package to message
    message = original.SerializeToString()

    # Parse the package
    new_pacakge = DriftPackage()
    new_pacakge.ParseFromString(message)
    print(f"Package ID={new_pacakge.id} type={MetaInfo.DataType.Name(new_pacakge.meta.type)}")

    payload = DataPayload()
    new_pacakge.data[0].Unpack(payload)

    buffer = WaveletBuffer.parse(payload.data)
    print(f"WaveletBuffer {buffer}")
